server/controllers/auth-controller.ts

```diff
@@ -1,61 +1,85 @@
-  const state = JSON.parse(
+function encodeLoginStateFromRequest(req: Request) {
+  return Buffer.from(
+    JSON.stringify({
+      returnTo: req.query.returnTo,
+      clientUrl: req.query.clientUrl,
+    })
+  ).toString('base64');
+}
+function decodeLoginStateFromRequest(req: Request) {
+  const parsed = JSON.parse(
     Buffer.from(req.query.state!.toString(), 'base64').toString()
   );
-  const returnTo = state['returnTo'] || '';
-  const clientUrl = state['clientUrl'];
+  return {
+    returnTo: parsed['returnTo'] as string,
+    clientUrl: parsed['clientUrl'] as string,
+  };
+}
+export async function refreshToken(
+  req: Request,
+  res: Response,
+  next: NextFunction
+) {
+  try {
+    const { userId, token } = req.body;
+    const newToken = await refreshAccessToken(userId, token);
+    res.json(newToken);
+  } catch (err) {
+    next(err);
+  }
+}
 
-  const user = JSON.stringify(req.user) || '';
+export function logout(req: Request, res: Response) {
+  req.logout(() => {
+    req.session.destroy(() => {
+      res.clearCookie('connect.sid');
+      res.redirect('/');
+    });
+  });
+}
 
-  res.redirect(
-    `${clientUrl}/logged-in?returnTo=${returnTo}&user=${encodeURIComponent(
-      user
-    )}`
-  );
-};
+export function googleLoginCallback(
+  req: Request,
+  res: Response,
+  next: NextFunction
+) {
+  log.debug('Google login called back.');
+  passport.authenticate('google', {
+    failureRedirect: '/api/auth/login',
+  })(req, res, next);
+}
 
-export const logout = (req: Request, res: Response) => {
-  req.logOut();
-  res.redirect('/');
-};
+export function googleLoginDone(
+  req: Request,
+  res: Response,
+  next: NextFunction
+) {
+  log.debug('Google login process done, redirecting client.');
+  const { returnTo, clientUrl } = decodeLoginStateFromRequest(req);
+  const userValue = encodeURIComponent(JSON.stringify(req.user) || '');
+  res.redirect(`${clientUrl}/logged-in?returnTo=${returnTo}&user=${userValue}`);
+}
 
-/**
- * Reissues access token based on userId and old expired token.
- *
- * @param {Request} req
- * @param {Response} res
- */
-export const renewToken = (req: Request, res: Response) => {
-  const userId = req.body.userId;
-  const oldToken = req.body.token;
-  refreshToken(userId, oldToken)
-    .then((newToken) => {
-      res.json(newToken);
-    })
-    .catch((reason) => {
-      res.status(400).json({ status: 'error', message: reason });
-    });
-};
+export function googleLogin(req: Request, res: Response, next: NextFunction) {
+  log.debug('Executing Google login.');
+  passport.authenticate('google', {
+    accessType: 'offline',
+    prompt: 'consent',
+    state: encodeLoginStateFromRequest(req),
+  })(req, res, next);
+}
+
+export function universalLogin(req: Request, res: Response) {
+  // Redirect to Google authentication
+  log.debug('Redirecting to Google login.');
+  res.redirect(
+    `/api/auth/google?returnTo=${req.query.returnTo}&clientUrl=${req.query.clientUrl}`
```

Comment:
Remove redirect to handle everything in `googleLogin`?
Also make `auth/google` the default `auth/login`.